### ISSUE TITLE
Fan Speed Control Not Responding to Temperature Changes

### ISSUE THREAD

**Issue by PetrKlassen** [sarzeaud/coreliquid_driver#2]
The program runs without errors and accurately detects the temperature (I debugged it). However, it does not seem to adjust the fan speeds based on the temperature readings.

OS: Arch Linux x86_64
Kernel: 6.11.6-arch1-1

**Comment by sarzeaud** [comment]
Hello,
Thanks for your interest.
The debugging you had to do is probably du to your os, which is not the same as mine (Debian). However, if you found a more universal way to detect the temperature, let me know.
Concerning the fan speed, which mode did you choose? Mode 4 does not change the speed. You should choose 0, 1, 2 or 5.

**Comment by PetrKlassen** [comment]
I try 2 and 5. the CPU temperature reaches 90 degree

**Comment by sarzeaud** [comment]
And you do not hear the fan turning higher?

**Comment by PetrKlassen** [comment]
nope, only case fan

**Comment by sarzeaud** [comment]
Ok, can you confirm that you started the daemon through systemctl ? Maybe a 
`systemctl | grep running `
could confirm it?

**Comment by PetrKlassen** [comment]
I try direct start with my additional temp log:
```
╰─❯ doas ./bin/out -M 5 startd
doas (petr@pc) password:
Current temp: 34
Current temp: 35
```

**Comment by sarzeaud** [comment]
You’re right, it should just work as well.
The last thing I can think of is usb communication. You should perhaps check that the handle returned by hid_open is not null and also control the hid_write return value. I don’t know if lsusb is available on arch Linux, but if so it could be worth checking that your aio is seen and has the right vid and pid.
A last thing, if you made several attempts to debug the program, it may be worth rebooting, as the usb management may be lost meanwhile.

**Comment by PetrKlassen** [comment]
no, nothing helps. count of written bites, device pid/vid are right. something strange


```
lsusb -vt
    |__ Port 013: Dev 008, If 0, Class=Hub, Driver=hub/4p, 480M
        ID 05e3:0608 Genesys Logic, Inc. Hub
        |__ Port 003: Dev 012, If 0, Class=Human Interface Device, Driver=usbhid, 480M
            ID 0db0:9ba6 Micro Star International
        |__ Port 004: Dev 014, If 0, Class=Vendor Specific Class, Driver=[none], 12M
            ID 0db0:6a05 Micro Star International
        |__ Port 004: Dev 014, If 1, Class=Human Interface Device, Driver=usbhid, 12M
            ID 0db0:6a05 Micro Star International
```

**Comment by sarzeaud** [comment]
Yes, very strange.
Perhaps could you ask some information to the device, using the protocol described here:
https://github.com/liquidctl/liquidctl/blob/main/docs/developer/protocol/coreliquid.md

**Comment by sarzeaud** [comment]
You can also try to display some information on the aio screen, to check if it reacts to some messages. You will find how to do this on the protocol page I just mentionned.

**Comment by sarzeaud** [comment]
I think the problem comes from the device having two interfaces on your system: it appears twice in the lsusb listing, with If=0 and If=1. When you open the device, I’m pretty sure it is the first one, with If=0, for which the driver must be MSI’s.
Have a look here to know how to choose the right device:
https://github.com/signal11/hidapi/issues/246

Not so sure finally, my configuration is the same (also give you the bus info line, just in case):

```
lsusb -vt
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/16p, 480M
    ID 1d6b:0002 Linux Foundation 2.0 root hub
    |__ Port 13: Dev 6, If 0, Class=Hub, Driver=hub/4p, 480M
        ID 05e3:0608 Genesys Logic, Inc. Hub
        |__ Port 2: Dev 12, If 0, Class=Vendor Specific Class, Driver=, 12M
            ID 0db0:6a05 Micro Star International 
        |__ Port 2: Dev 12, If 1, Class=Human Interface Device, Driver=usbhid, 12M
            ID 0db0:6a05 Micro Star International 
        |__ Port 1: Dev 10, If 0, Class=Human Interface Device, Driver=usbhid, 480M
            ID 0db0:9ba6 Micro Star International 
```

**Comment by PetrKlassen** [comment]
I tested some additional options, turned off the other case fans, and realized that the fans do indeed speed up. However, the CPU cooling still performs worse than on Windows (On Windows, the CPU temperature never reached 90 degrees, but here it does during demanding tasks.). It's hard to say what this is related to. I will continue to investigate and try passing additional parameters, such as the GPU temperature and CPU clock speed. It feels like the airflow from the fans is weaker than it should be at maximum speed.

It would be ironic if the CPU clock speed is simply throttled on Windows (I have an Intel i9 12900).

**Comment by sarzeaud** [comment]
That’s good news.
When you compare to windows, what do you use to get the cpu temperature? If it is the MSI tool, you do not know how they get a unique temperature. Your cpu has a temperature of each core, and there is a possibility that they compute a kind of mean temperature. It also can be filtered to avoid jumps in fan speed.
However, what you can be sure of is that the aio only uses one cpu temperature to choose the fan speed. Gpu temperature or cpu clock speed are not used for this but only for the display.
Another thing you can suspect is that windows does silly things, just like it is supposed to do.
